## Confirming changes with updated tests

Following up on this PR, I'm on the verge of pushing up changes to tests to cover this and want to demonstrate that these tests fail on `trunk` before demonstrating they pass on the branch:

```shell
Running tool: /opt/homebrew/bin/go test -timeout 30s -run ^TestNewCmdListBaseRepoFuncs$ github.com/cli/cli/v2/pkg/cmd/secret/list

--- FAIL: TestNewCmdListBaseRepoFuncs (0.00s)
    --- FAIL: TestNewCmdListBaseRepoFuncs/when_GH_REPO_env_var_is_provided,_the_factory_base_repo_func_is_used (0.00s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x2 addr=0x28 pc=0x1032fdec0]

goroutine 6 [running]:
testing.tRunner.func1.2({0x10358c3a0, 0x1039e9a50})
	/opt/homebrew/Cellar/go/1.24.0/libexec/src/testing/testing.go:1734 +0x1ac
testing.tRunner.func1()
	/opt/homebrew/Cellar/go/1.24.0/libexec/src/testing/testing.go:1737 +0x334
panic({0x10358c3a0?, 0x1039e9a50?})
	/opt/homebrew/Cellar/go/1.24.0/libexec/src/runtime/panic.go:787 +0x124
github.com/cli/cli/v2/internal/ghrepo.IsSame({0x10362d828, 0x140001da7b0}, {0x0, 0x0})
	/Users/andyfeller/Documents/workspace/cli/cli-10539/internal/ghrepo/repo.go:80 +0x40
github.com/cli/cli/v2/pkg/cmd/secret/list.TestNewCmdListBaseRepoFuncs.func2(0x140001ea380)
	/Users/andyfeller/Documents/workspace/cli/cli-10539/pkg/cmd/secret/list/list_test.go:217 +0x3f8
testing.tRunner(0x140001ea380, 0x140001d37a0)
	/opt/homebrew/Cellar/go/1.24.0/libexec/src/testing/testing.go:1792 +0xe4
created by testing.(*T).Run in goroutine 4
	/opt/homebrew/Cellar/go/1.24.0/libexec/src/testing/testing.go:1851 +0x374
FAIL	github.com/cli/cli/v2/pkg/cmd/secret/list	0.208s
FAIL


Running tool: /opt/homebrew/bin/go test -timeout 30s -run ^TestNewCmdDeleteBaseRepoFuncs$ github.com/cli/cli/v2/pkg/cmd/secret/delete

--- FAIL: TestNewCmdDeleteBaseRepoFuncs (0.00s)
    --- FAIL: TestNewCmdDeleteBaseRepoFuncs/when_GH_REPO_env_var_is_provided,_the_factory_base_repo_func_is_used (0.00s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x2 addr=0x28 pc=0x104758000]

goroutine 37 [running]:
testing.tRunner.func1.2({0x104993fa0, 0x104dac9d0})
	/opt/homebrew/Cellar/go/1.24.0/libexec/src/testing/testing.go:1734 +0x1ac
testing.tRunner.func1()
	/opt/homebrew/Cellar/go/1.24.0/libexec/src/testing/testing.go:1737 +0x334
panic({0x104993fa0?, 0x104dac9d0?})
	/opt/homebrew/Cellar/go/1.24.0/libexec/src/runtime/panic.go:787 +0x124
github.com/cli/cli/v2/internal/ghrepo.IsSame({0x104a30788, 0x140001fe6f0}, {0x0, 0x0})
	/Users/andyfeller/Documents/workspace/cli/cli-10539/internal/ghrepo/repo.go:80 +0x40
github.com/cli/cli/v2/pkg/cmd/secret/delete.TestNewCmdDeleteBaseRepoFuncs.func2(0x14000129180)
	/Users/andyfeller/Documents/workspace/cli/cli-10539/pkg/cmd/secret/delete/delete_test.go:238 +0x3f8
testing.tRunner(0x14000129180, 0x140001f3810)
	/opt/homebrew/Cellar/go/1.24.0/libexec/src/testing/testing.go:1792 +0xe4
created by testing.(*T).Run in goroutine 35
	/opt/homebrew/Cellar/go/1.24.0/libexec/src/testing/testing.go:1851 +0x374
FAIL	github.com/cli/cli/v2/pkg/cmd/secret/delete	0.253s
FAIL


Running tool: /opt/homebrew/bin/go test -timeout 30s -run ^TestNewCmdSetBaseRepoFuncs$ github.com/cli/cli/v2/pkg/cmd/secret/set

--- FAIL: TestNewCmdSetBaseRepoFuncs (0.00s)
    --- FAIL: TestNewCmdSetBaseRepoFuncs/when_GH_REPO_env_var_is_provided,_the_factory_base_repo_func_is_used (0.00s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x2 addr=0x28 pc=0x100ac8a10]

goroutine 6 [running]:
testing.tRunner.func1.2({0x100d14aa0, 0x101138b40})
	/opt/homebrew/Cellar/go/1.24.0/libexec/src/testing/testing.go:1734 +0x1ac
testing.tRunner.func1()
	/opt/homebrew/Cellar/go/1.24.0/libexec/src/testing/testing.go:1737 +0x334
panic({0x100d14aa0?, 0x101138b40?})
	/opt/homebrew/Cellar/go/1.24.0/libexec/src/runtime/panic.go:787 +0x124
github.com/cli/cli/v2/internal/ghrepo.IsSame({0x100db2cb8, 0x140001d8810}, {0x0, 0x0})
	/Users/andyfeller/Documents/workspace/cli/cli-10539/internal/ghrepo/repo.go:80 +0x40
github.com/cli/cli/v2/pkg/cmd/secret/set.TestNewCmdSetBaseRepoFuncs.func2(0x140001ea380)
	/Users/andyfeller/Documents/workspace/cli/cli-10539/pkg/cmd/secret/set/set_test.go:337 +0x3f8
testing.tRunner(0x140001ea380, 0x140001e22a0)
	/opt/homebrew/Cellar/go/1.24.0/libexec/src/testing/testing.go:1792 +0xe4
created by testing.(*T).Run in goroutine 4
	/opt/homebrew/Cellar/go/1.24.0/libexec/src/testing/testing.go:1851 +0x374
FAIL	github.com/cli/cli/v2/pkg/cmd/secret/set	0.248s
FAIL
```

--- a/acceptance/testdata/secret/secret-require-remote-disambiguation.txtar
+++ b/acceptance/testdata/secret/secret-require-remote-disambiguation.txtar
@@ -48,12 +48,15 @@ stdout 'TEST_SECRET_NAME'
 # Secret delete using --repo flag does not require disambiguation
 exec gh secret delete 'TEST_SECRET_NAME' --repo ${ORG}/${REPO}-fork
 
+# Setup GH_REPO for testing environment variable behavior
+env GH_REPO=${ORG}/${REPO}-fork
+
 # Secret set using GH_REPO env var does not require disambiguation
-exec gh secret set 'TEST_SECRET_NAME2' --body 'TEST_SECRET_VALUE2' --repo ${ORG}/${REPO}-fork
+exec gh secret set 'TEST_SECRET_NAME2' --body 'TEST_SECRET_VALUE2'
 
 # Secret list using --repo flag does not require disambiguation
-exec gh secret list --repo ${ORG}/${REPO}-fork
+exec gh secret list
 stdout 'TEST_SECRET_NAME2'
 
 # Secret delete using --repo flag does not require disambiguation
-exec gh secret delete 'TEST_SECRET_NAME2' --repo ${ORG}/${REPO}-fork
+exec gh secret delete 'TEST_SECRET_NAME2'

--- a/acceptance/testdata/secret/secret-require-remote-disambiguation.txtar
+++ b/acceptance/testdata/secret/secret-require-remote-disambiguation.txtar
@@ -34,3 +34,26 @@ stderr 'multiple remotes detected. please specify which repo to use by providing
 # Secret delete requires disambiguation
 ! exec gh secret delete 'TEST_SECRET_NAME'
 stderr 'multiple remotes detected. please specify which repo to use by providing the -R, --repo argument'
+
+# Move out of the fork repo to test whether secret commands work without local repository context
+cd ..
+
+# Secret set using --repo flag does not require disambiguation
+exec gh secret set 'TEST_SECRET_NAME' --body 'TEST_SECRET_VALUE' --repo ${ORG}/${REPO}-fork
+
+# Secret list using --repo flag does not require disambiguation
+exec gh secret list --repo ${ORG}/${REPO}-fork
+stdout 'TEST_SECRET_NAME'
+
+# Secret delete using --repo flag does not require disambiguation
+exec gh secret delete 'TEST_SECRET_NAME' --repo ${ORG}/${REPO}-fork
+
+# Secret set using GH_REPO env var does not require disambiguation
+exec gh secret set 'TEST_SECRET_NAME2' --body 'TEST_SECRET_VALUE2' --repo ${ORG}/${REPO}-fork
+
+# Secret list using --repo flag does not require disambiguation
+exec gh secret list --repo ${ORG}/${REPO}-fork
+stdout 'TEST_SECRET_NAME2'
+
+# Secret delete using --repo flag does not require disambiguation
+exec gh secret delete 'TEST_SECRET_NAME2' --repo ${ORG}/${REPO}-fork

--- a/pkg/cmd/secret/delete/delete_test.go
+++ b/pkg/cmd/secret/delete/delete_test.go
@@ -144,6 +144,7 @@ func TestNewCmdDeleteBaseRepoFuncs(t *testing.T) {
 	tests := []struct {
 		name          string
 		args          string
+		env           map[string]string
 		prompterStubs func(*prompter.MockPrompter)
 		wantRepo      ghrepo.Interface
 		wantErr       error
@@ -154,14 +155,22 @@ func TestNewCmdDeleteBaseRepoFuncs(t *testing.T) {
 			wantRepo: ghrepo.New("owner", "repo"),
 		},
 		{
-			name: "when there is no repo flag provided, and no prompting, the base func requiring no ambiguity is used",
+			name: "when GH_REPO env var is provided, the factory base repo func is used",
+			args: "SECRET_NAME",
+			env: map[string]string{
+				"GH_REPO": "owner/repo",
+			},
+			wantRepo: ghrepo.New("owner", "repo"),
+		},
+		{
+			name: "when there is no repo flag or GH_REPO env var provided, and no prompting, the base func requiring no ambiguity is used",
 			args: "SECRET_NAME",
 			wantErr: shared.AmbiguousBaseRepoError{
 				Remotes: remotes,
 			},
 		},
 		{
-			name: "when there is no repo flag provided, and can prompt, the base func resolving ambiguity is used",
+			name: "when there is no repo flag or GH_REPO env var provided, and can prompt, the base func resolving ambiguity is used",
 			args: "SECRET_NAME",
 			prompterStubs: func(pm *prompter.MockPrompter) {
 				pm.RegisterSelect(
@@ -197,6 +206,10 @@ func TestNewCmdDeleteBaseRepoFuncs(t *testing.T) {
 				Remotes: func() (ghContext.Remotes, error) {
 					return remotes, nil
 				},
+			}
+
+			for k, v := range tt.env {
+				t.Setenv(k, v)
 			}
 
 			argv, err := shlex.Split(tt.args)

--- a/pkg/cmd/secret/list/list_test.go
+++ b/pkg/cmd/secret/list/list_test.go
@@ -124,6 +124,7 @@ func TestNewCmdListBaseRepoFuncs(t *testing.T) {
 	tests := []struct {
 		name          string
 		args          string
+		env           map[string]string
 		prompterStubs func(*prompter.MockPrompter)
 		wantRepo      ghrepo.Interface
 		wantErr       error
@@ -134,14 +135,21 @@ func TestNewCmdListBaseRepoFuncs(t *testing.T) {
 			wantRepo: ghrepo.New("owner", "repo"),
 		},
 		{
-			name: "when there is no repo flag provided, and no prompting, the base func requiring no ambiguity is used",
+			name: "when GH_REPO env var is provided, the factory base repo func is used",
+			env: map[string]string{
+				"GH_REPO": "owner/repo",
+			},
+			wantRepo: ghrepo.New("owner", "repo"),
+		},
+		{
+			name: "when there is no repo flag or GH_REPO env var provided, and no prompting, the base func requiring no ambiguity is used",
 			args: "",
 			wantErr: shared.AmbiguousBaseRepoError{
 				Remotes: remotes,
 			},
 		},
 		{
-			name: "when there is no repo flag provided, and can prompt, the base func resolving ambiguity is used",
+			name: "when there is no repo flag or GH_REPO env var provided, and can prompt, the base func resolving ambiguity is used",
 			args: "",
 			prompterStubs: func(pm *prompter.MockPrompter) {
 				pm.RegisterSelect(
@@ -177,6 +185,10 @@ func TestNewCmdListBaseRepoFuncs(t *testing.T) {
 				Remotes: func() (ghContext.Remotes, error) {
 					return remotes, nil
 				},
+			}
+
+			for k, v := range tt.env {
+				t.Setenv(k, v)
 			}
 
 			argv, err := shlex.Split(tt.args)

--- a/pkg/cmd/secret/set/set_test.go
+++ b/pkg/cmd/secret/set/set_test.go
@@ -243,6 +243,7 @@ func TestNewCmdSetBaseRepoFuncs(t *testing.T) {
 	tests := []struct {
 		name          string
 		args          string
+		env           map[string]string
 		prompterStubs func(*prompter.MockPrompter)
 		wantRepo      ghrepo.Interface
 		wantErr       error
@@ -253,14 +254,22 @@ func TestNewCmdSetBaseRepoFuncs(t *testing.T) {
 			wantRepo: ghrepo.New("owner", "repo"),
 		},
 		{
-			name: "when there is no repo flag provided, and no prompting, the base func requiring no ambiguity is used",
+			name: "when GH_REPO env var is provided, the factory base repo func is used",
+			args: "SECRET_NAME",
+			env: map[string]string{
+				"GH_REPO": "owner/repo",
+			},
+			wantRepo: ghrepo.New("owner", "repo"),
+		},
+		{
+			name: "when there is no repo flag or GH_REPO env var provided, and no prompting, the base func requiring no ambiguity is used",
 			args: "SECRET_NAME",
 			wantErr: shared.AmbiguousBaseRepoError{
 				Remotes: remotes,
 			},
 		},
 		{
-			name: "when there is no repo flag provided, and can prompt, the base func resolving ambiguity is used",
+			name: "when there is no repo flag or GH_REPO env var provided, and can prompt, the base func resolving ambiguity is used",
 			args: "SECRET_NAME",
 			prompterStubs: func(pm *prompter.MockPrompter) {
 				pm.RegisterSelect(
@@ -296,6 +305,10 @@ func TestNewCmdSetBaseRepoFuncs(t *testing.T) {
 				Remotes: func() (ghContext.Remotes, error) {
 					return remotes, nil
 				},
+			}
+
+			for k, v := range tt.env {
+				t.Setenv(k, v)
 			}
 
 			argv, err := shlex.Split(tt.args)

